### PR TITLE
QueryWrapper Refactor

### PR DESCRIPTION
**Pull request by nickgros**
* Converted QueryWrapper from one class component to two function components with contexts
  * QueryWrapper now creates a QueryContext, which manages query data
  * QueryVisualizationWrapper creates a QueryVisualizationContext, which manages visualization state
  * QueryWrapper now uses react-query for managing fetched data
* Updated CardContainerLogic to use QueryWrapper
* Updated all components that relied on QueryWrapper to use context
* Removed StackedBarChart. It is still referenced in portals configurations, but is not actually shown anywhere.
* Corresponding test fixes for all above changes

### CODE REVIEW

File: src/lib/containers/QueryWrapper.tsx [hunk 1]
```diff
@@ -99,140 +127,127 @@ export type LockedFacet = {
...
+    // We use `keepPreviousData` because we don't want to clear out the current data when the query is modified via the UI
+    keepPreviousData: true,
+  })
+
+  // Indicate if we're fetching data for the first time (queryIsLoading) or if we're fetching data for a brand new query (isPreviousData)
+  const isLoadingNewBundle = queryIsLoading || isPreviousData
+
+  const entityId = parseEntityIdFromSqlStatement(lastQueryRequest.query.sql)
+
+  const { data: entity } = useGetEntity(entityId)
```
Comment (nickgros) [inline src/lib/containers/QueryWrapper.tsx:159]: Some components use the Entity object, e.g. SynapseTable needs it to see if it's a File View. We will just always fetch it and include it in the QueryContext.

File: src/lib/utils/SynapseClient.ts [hunk 2]
```diff
@@ -220,65 +235,50 @@ export type SynapseClientError = {
...
+/**
+ * Fetches data, retrying if the HTTP status code indicates that it could be retried. Contains custom logic for
+ * handling errors returned by the Synapse backend.
+ * @throws SynapseClientError
+ */
+const fetchWithExponentialTimeout = async <TResponse>(
   url: RequestInfo,
   options: RequestInit,
   delayMs = 1000,
-): Promise<T> => {
```
Comment (nickgros) [inline src/lib/utils/SynapseClient.ts:227]: I personally could not reason with this callback hell, so I rewrote this

File: src/lib/utils/SynapseClient.ts [hunk 3]
```diff
@@ -220,65 +235,50 @@ export type SynapseClientError = {
...
-      } else {
-        return Promise.reject(error)
-      }
-    })
+): Promise<TResponse> => {
+  let response = await fetch(url, options)
+  while (response.status && RETRY_STATUS_CODES.includes(response.status)) {
+    await delay(delayMs)
+    response = await fetch(url, options)
+  }
```
Comment (nickgros) [inline src/lib/utils/SynapseClient.ts:252]: Retries are now in a while loop instead of recursively called.

File: src/lib/utils/SynapseClient.ts [hunk 4]
```diff
@@ -497,39 +458,74 @@ export const getFileHandleByIdURL = (
...
+     * While we technically already have the failure reason in the response, the HTTP response doesn't give a helpful error code (e.g. 403)
+     * that we can use for an error banner. We can get the HTTP code if we fetch the response body directly.
+     */
+    doGet<TResponse>(
+      responseBodyEndpoint,
+      accessToken,
+      undefined,
+      BackendDestinationEnum.REPO_ENDPOINT,
+    )
+  }
```
Comment (nickgros) [inline src/lib/utils/SynapseClient.ts:506]: I tried to make all requests use "/repo/v1/asynchronous/job/${jobId}" before realizing that we no longer get the HTTP status code for errors, which tells our ErrorBanner what to do. So if the async job fails, fetch it one last time and get the thrown error.

File: Examples.md [hunk 5]
```diff
@@ -1,248 +0,0 @@
-# Examples
```
Comment (nickgros) [inline Examples.md:1]: Most of this was incorrect. The Styleguidist demo page makes this obsolete.

File: src/lib/containers/QueryWrapper.tsx [hunk 6]
```diff
@@ -99,140 +127,133 @@ export type LockedFacet = {
...
+  const {
+    data: infiniteData,
+    hasNextPage,
+    fetchPreviousPage,
+    fetchNextPage,
+    isFetchingNextPage,
+    isLoading: queryIsLoading,
+    refetch,
+    error,
+    isPreviousData,
```
Comment (nickgros) [inline src/lib/containers/QueryWrapper.tsx:147]: Rename for clarity to show that we're in between new queries

File: src/lib/containers/CardContainerLogic.tsx [hunk 7]
```diff
@@ -97,196 +86,59 @@ export type CardConfiguration = {
...
+    <QueryWrapper {...props} initQueryRequest={initQueryRequest}>
+      <QueryVisualizationWrapper
+        rgbIndex={props.rgbIndex}
+        unitDescription={props.unitDescription}
+        facetAliases={props.facetAliases}
+      >
+        <CardContainer {...props} />
+        <QueryContextConsumer>
+          {queryContext => <ErrorBanner error={queryContext?.error} />}
+        </QueryContextConsumer>
```
Comment (nickgros) [inline src/lib/containers/CardContainerLogic.tsx:138]: Pull out into QueryWrapperErrorBanner

File: src/lib/containers/QueryVisualizationWrapper.tsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+/*
+  For details page: to lock a facet name (e.g. study, grant) so that the facet name
+  and its all possible values will not appear on the details page. The facet name is
+  given by the url's search param. The type is defined here so that other child components
+  in SRC won't generate type errors.
+ */
+export type LockedFacet = {
+  facet?: string
+  value?: string
+}
```
Comment (nickgros) [inline src/lib/containers/QueryVisualizationWrapper.tsx:89]: might be duplicated

File: src/lib/utils/hooks/SynapseAPI/useGetQueryResultBundle.ts [hunk 9]
```diff
@@ -1,22 +1,126 @@
...
+        if (firstPage.jobState !== 'COMPLETE') {
+          return undefined
+        }
+        const request = firstPage.requestBody
+        if (request.query.offset == null || request.query.offset === 0) {
+          return undefined
+        }
+        const pageSize = request.query.limit ?? DEFAULT_PAGE_SIZE
+
+        return request.query.offset - pageSize
```
Comment (nickgros) [inline src/lib/utils/hooks/SynapseAPI/useGetQueryResultBundle.ts:104]: Set floor of 0

File: src/lib/utils/SynapseClient.ts [hunk 10]
```diff
@@ -220,65 +235,50 @@ export type SynapseClientError = {
...
-          return fetchWithExponentialTimeout<T>(url, options, delayMs * 2)
-        })
-      } else {
-        return Promise.reject(error)
-      }
-    })
+): Promise<TResponse> => {
+  let response = await fetch(url, options)
+  while (response.status && RETRY_STATUS_CODES.includes(response.status)) {
+    await delay(delayMs)
```
Comment (nickgros) [inline src/lib/utils/SynapseClient.ts:250]: Make this delay exponential

File: src/lib/utils/functions/EntityTypeUtils.ts [hunk 11]
```diff
@@ -132,3 +134,13 @@ export function isVersionableEntityType(type: EntityType): boolean {
...
 }
+
+export function isEntityView(entity: Entity): entity is EntityView {
+  return (
+    entity.concreteType === 'org.sagebionetworks.repo.model.table.EntityView'
+  )
+}
+
+export function hasFilesInView(entity: EntityView) {
+  return (entity.viewTypeMask & 0x01) != 0
```
Comment (nickgros) [inline src/lib/utils/functions/EntityTypeUtils.ts:145]: replace 0x01 with constant

File: src/lib/utils/functions/sqlFunctions.ts [hunk 12]
```diff
@@ -178,6 +178,13 @@ export const parseEntityIdFromSqlStatement = (sql: string): string => {
   return matches && matches[0] ? matches[0].substr(5).trim() : ''
 }
 
+export const parseEntityIdAndVersionFromSqlStatement = (
+  sql: string,
+): string => {
+  const matches = sql.match(/(from)\s+(syn)\d+(\.\d+)?/gi)
+  return matches && matches[0] ? matches[0].substr(5).trim() : ''
+}
+
```
Comment (nickgros) [inline src/lib/utils/functions/sqlFunctions.ts:187]: Remove this, unused.

File: src/lib/containers/table/TopLevelControls.tsx [hunk 13]
```diff
@@ -78,58 +74,60 @@ const controls: Control[] = [
...
-  const [isFileView, setIsFileView] = useState(false)
+    setColumnsToShowInTable,
+  } = useQueryVisualizationContext()
+
+  const isFileView =
+    entity?.concreteType ===
+      'org.sagebionetworks.repo.model.table.EntityView' &&
+    (entity as EntityView | undefined)?.viewTypeMask === 1
+  const isDataset =
+    entity?.concreteType === 'org.sagebionetworks.repo.model.table.Dataset'
```
Comment (nickgros) [inline src/lib/containers/table/TopLevelControls.tsx:111]: Partial change that I should remove

File: src/lib/containers/table/TopLevelControls.tsx [hunk 14]
```diff
@@ -78,58 +74,60 @@ const controls: Control[] = [
...
 
-  useEffect(() => {
-    const getIsFileView = async () => {
-      const entityData = await SynapseClient.getEntity(accessToken, entityId)
-      setIsFileView(entityData.concreteType.includes('EntityView'))
-    }
-    getIsFileView()
-  }, [entityId, accessToken])
-
   const refresh = () => {
```
Comment (nickgros) [inline src/lib/containers/table/TopLevelControls.tsx:129]: Still needed?

File: src/lib/containers/table/TopLevelControls.tsx [hunk 15]
```diff
@@ -138,40 +136,65 @@ const TopLevelControls = (
...
+                  {!hideQueryCount && (
+                    <QueryCount
+                      sql={getInitQueryRequest().query.sql}
+                      parens={true}
+                    />
+                  )}
+                </Typography>
+              )}
+            </div>
+          )}
```
Comment (nickgros) [inline src/lib/containers/table/TopLevelControls.tsx:187]: Remove

File: src/lib/containers/table/TopLevelControls.tsx [hunk 16]
```diff
@@ -196,11 +219,12 @@ const TopLevelControls = (
             } else if (key === 'showDownloadConfirmation') {
               return (
                 <DownloadOptions
+                  key={key}
                   darkTheme={true}
                   onDownloadFiles={() => setControlState(key)}
                   queryResultBundle={data}
-                  queryBundleRequest={getLastQueryRequest!()}
-                  isFileView={isFileView && !hideDownload}
```
Comment (nickgros) [inline src/lib/containers/table/TopLevelControls.tsx:203]: Reason for old code?

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**